packages/dds/quorum/src/quorum.ts

```diff
@@ -0,0 +1,373 @@
+
+    /**
+     * Constructs a new quorum. If the object is non-local an id and service interfaces will
+     * be provided
+     *
+     * @param runtime - data store runtime the quorum belongs to
+     * @param id - optional name of the quorum
+     */
+    constructor(id: string, runtime: IFluidDataStoreRuntime, attributes: IChannelAttributes) {
+        super(id, runtime, attributes);
+
+        this.incomingOp.on("set", this.handleIncomingSet);
+        this.incomingOp.on("delete", this.handleIncomingDelete);
+        this.incomingOp.on("noop", this.handleIncomingNoOp);
+
+        this.disconnectWatcher.on("disconnect", () => {
+            // assert(this.runtime.clientId !== undefined, 0x1d3 /* "Missing client id on disconnect" */);
+            // TODO Handle appropriately
+        });
+    }
+
+    public has(key: string): boolean {
+        return this.values.get(key)?.accepted !== undefined;
+    }
+
+    public get(key: string): any {
+        return this.values.get(key)?.accepted?.value;
+    }
+
+    // TODO: Should this return differently for a value of undefined vs. a pending delete?
+    public getPending(key: string): any {
+        const pending = this.values.get(key)?.pending;
+        if (pending === undefined || pending.type === "delete") {
+            return undefined;
+        }
+
+        return pending.value;
+    }
+
+    public set(key: string, value: any): void {
+        // TODO: handle detached scenario, just auto accept basically
+
+        const setOp: IQuorumSetOperation = {
+            type: "set",
+            key,
+            value,
+            refSeq: this.runtime.deltaManager.lastSequenceNumber,
+        };
+
+        this.submitLocalMessage(setOp);
+    }
+
+    public delete(key: string): void {
+        // TODO: handle detached scenario, just auto accept basically
+
+        const deleteOp: IQuorumDeleteOperation = {
+            type: "delete",
+            key,
+            refSeq: this.runtime.deltaManager.lastSequenceNumber,
+        };
+
+        this.submitLocalMessage(deleteOp);
+    }
+
+    private readonly handleIncomingSet = (key: string, value: any, refSeq: number, setSequenceNumber: number) => {
+        const currentValue = this.values.get(key);
+        // A proposal is valid if the value is unknown
+        // or if it was made with knowledge of the most recently accepted value
+        const proposalValid =
+            currentValue === undefined
+            || (currentValue.pending === undefined && currentValue.accepted.sequenceNumber <= refSeq);
+        if (!proposalValid) {
+            // Drop invalid proposals on the ground.  If set() returns a promise we will need to resolve it though.
+            return;
+        }
+
+        const accepted = currentValue?.accepted;
+
+        this.values.set(key, { accepted, pending: { type: "set", value, sequenceNumber: setSequenceNumber }});
+
+        this.emit("pending", key);
```

Comment:
looks like we don't support a reject semantic here, but if we ever will we would want to support async here.

Reply:
I've thought about this a little bit - I think the first approach I would explore might look something like:

```typescript
(event: "pending", listener: (
    key: string,
    requestDelay: () => void,
    accept: () => void,
    reject: () => void,
) => void);
```

Where the `requestDelay()` callback sends an op synchronously to add a further condition to acceptance -- that the clientId that sent the request either accepts the request (via `accept()`) or otherwise leaves the quorum.

This still forces an async response so we can continue to use MSN as the driving force for acceptance, but permits async decision-making for connected clients.

Reply:
I call it out below, but it will be tricky to do coordination vis the MSN. Even in this example simply sending the `requestDelay` op could send the msn forward. in this case you can probably have logic to not accept in that case, but it adds complexity.

Reply:
With the new explicit ack approach this becomes even simpler, since the clause is no longer a compound of MSN + delaying clients -- no longer need a `requestDelay` op.